### PR TITLE
Syntax changes for recursive pattern-matching feature

### LINKED ISSUE #32291: Rename deconstruct_pattern to positional_pattern
We should change both the spec grammar and the syntax nodes.

### LINKED ISSUE #32292: Permit a trailing comma inside a switch expression
Per LDM 2019-01-09 we have accepted the proposal in https://github.com/dotnet/csharplang/issues/2098 to allow a trailing comma inside a switch expression
``` c#
_ = e switch
    {
        1 => 2,
        3 => 4, // comma now allowed here
    }
```
Similarly, we have accepted a proposal to allow a trailing comma inside a property subpattern.
``` c#
_ = e is T
    {
        X: 2,
        Y: 3, // comma now allowed here
    };
```
This needs to be implemented.

### PR DESCRIPTION
**Pull request by gafter**
Rename "deconstruct pattern" to "positional pattern" in the APIs.
Fixes #32291

Permit a trailing commas in two places
- Permit a trailing comma after the last arm of a switch expression
- Permit a trailing comma after the last subpattern of a property pattern clause

Fixes #32292

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs [hunk 1]
```diff
@@ -137,11 +137,11 @@ private bool LooksLikeTypeOfPattern(SyntaxKind tk)
         //
         // Parse an expression where a declaration expression would be permitted. This is suitable for use after
         // the `out` keyword in an argument list, or in the elements of a tuple literal (because they may
-        // be on the left-hand-side of a deconstruction). The first element of a tuple is handled slightly
+        // be on the left-hand-side of a positional). The first element of a tuple is handled slightly
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs]: incomplete sentence?

File: src/Compilers/CSharp/Test/Syntax/Parsing/ParsingTests.cs [hunk 2]
```diff
@@ -1,6 +1,6 @@
 ﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
-//#define PARSING_TESTS_DUMP
+#define PARSING_TESTS_DUMP
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ParsingTests.cs]: Should be reverted

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests4.cs [hunk 3]
```diff
@@ -1718,15 +1718,15 @@ static int M1(bool? b1, bool? b2)
...
+            (null, true) => 6,
             };
     }
 }
 ";
             var compilation = CreatePatternCompilation(source);
             compilation.VerifyDiagnostics(
                 // (13,13): error CS8510: The pattern has already been handled by a previous arm of the switch expression.
-                //             (null, true) => 6
+                //             (null, true) => 6,
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests4.cs]: Consider adding a test with two trailing commas. (same feedback also applies to property pattern)

**Comment by jcouv** [comment]
Sorry, I failed to notice the "personal" label.

**Comment by gafter** [comment]
@dotnet/roslyn-compiler May I please have a couple of reviews of this set of syntax changes for pattern-matching?  It is targeting P3.

**Pull request merged**